File: src/db/schema.ts
import { boolean, date, integer, pgTable, text, timestamp, uuid, varchar } from "drizzle-orm/pg-core";
import { relations } from "drizzle-orm";


// This is the schema for the users table
export const users = pgTable("users", {

  // Primary key
  id: varchar("id", { length: 50 }).primaryKey().notNull(),

  // User fields
  userName: varchar("username", { length: 50 }).unique().notNull(),
  email: varchar("email", { length: 255 }).unique().notNull(),
  password: varchar("password", { length: 255 }),

  // Verification fields
  isAdmin: boolean("is_admin").default(false).notNull(),

  // timestamps
  createdAt: timestamp("created_at").defaultNow().notNull(),
  updatedAt: timestamp("updated_at")
    .defaultNow()
    .$onUpdate(() => new Date())
    .notNull(),
});


// This is the schema for the user_info table
// This table contains additional information about the user
export const userInfo = pgTable("user_info", {

  // Primary key
  id: uuid("id").defaultRandom().primaryKey().notNull(),

  // Foreign key
  // This is a reference to the users table
  userId: varchar("user_id", { length: 50 }).unique().notNull().references(() => users.id),

  // User info fields
  firstName: varchar("first_name", { length: 50 }).notNull(),
  lastName: varchar("last_name", { length: 50 }),

  // Optional fields
  bio: text("bio"),
  profilePic: varchar("profile_picture", { length: 255 }),
  dob: date("dob"),
  country: varchar("country", { length: 50 }),

  // timestamps
  createdAt: timestamp("created_at").defaultNow().notNull(),
  updatedAt: timestamp("updated_at")
    .defaultNow()
    .$onUpdate(() => new Date())
    .notNull(),
})

export const userRelations = relations(users, ({ one }) => ({
  userInfo: one(userInfo),
}));



// 

export type User = typeof users.$inferSelect;
export type UserInsert = typeof users.$inferInsert;

export type UserInfo = typeof userInfo.$inferSelect;
export type UserInfoInsert = typeof userInfo.$inferInsert;

File: schema.rb
class CreateUsersAndUserInfos < ActiveRecord::Migration[7.0]
  def change
    create_table :users, id: :uuid do |t|
      t.string  :user_name,         null: false, limit: 50,  index: { unique: true }
      t.string  :email,             null: false, limit: 255, index: { unique: true }
      t.string  :password,                          limit: 50
      t.boolean :is_admin,          null: false, default: false
      t.boolean :is_verified,       null: false, default: false
      t.string  :verification_code,               limit: 6
      t.timestamps
    end

    create_table :user_infos, id: :uuid do |t|
      t.uuid    :user_id,          null: false, index: { unique: true }
      t.string  :first_name,        null: false, limit: 50
      t.string  :last_name,                    limit: 50
      t.text    :bio
      t.string  :profile_picture,              limit: 255
      t.date    :dob
      t.string  :country,                   limit: 50
      t.timestamps
    end

    add_foreign_key :user_infos, :users, column: :user_id, primary_key: :id
  end
end

# app/models/user.rb
class User < ApplicationRecord
  has_one :user_info, dependent: :destroy
end

# app/models/user_info.rb
class UserInfo < ApplicationRecord
  belongs_to :user
end
